demo-multiroot-react-18/src/MultiRootEditorRichDemo.tsx

```diff
@@ -0,0 +1,211 @@
+
+		if ( isReadOnly ) {
+			disabledRoots.delete( root.rootName );
+			editor!.enableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		} else {
+			disabledRoots.add( root.rootName );
+			editor!.disableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		}
+
+		setDisabledRoots( new Set( disabledRoots ) );
+	};
+
+	// Function to simulate an error in the editor.
+	// It is used for testing purposes to trigger the Watchdog to restart the editor.
+	// Remove it in the actual integration.
+	const simulateError = () => {
+		setTimeout( () => {
+			const err: any = new Error( 'foo' );
+
+			err.context = editor;
+			err.is = () => true;
+
+			throw err;
+		} );
+	};
+
+	const addRoot = ( newRootAttributes: Record<string, unknown>, rootId?: string ) => {
+		const id = rootId || new Date().getTime();
+
+		for ( let i = 1; i <= numberOfRoots; i++ ) {
+			const rootName = `root-${ i }-${ id }`;
+
+			data[ rootName ] = '';
+
+			// Remove code related to rows if you don't need to handle multiple roots in one row.
+			attributes[ rootName ] = { ...newRootAttributes, order: i * 10, row: id };
+		}
+
+		setData( { ...data } );
+		setAttributes( { ...attributes } );
+		// Reset the <input> element to the default value.
+		setNumberOfRoots( 1 );
+	};
+
+	const removeRoot = ( rootName: string ) => {
+		setData( previousData => {
+			const { [ rootName! ]: _, ...newData } = previousData;
+
+			return { ...newData };
+		} );
+
+		setSelectedRoot( '' );
+	};
+
+	// Group elements based on their row attribute and sort them by order attribute.
+	// Grouping in a row is used for presentation purposes, and you may remove it in actual integration.
+	// However, we recommend ordering the roots, so that rows are put in a correct places when undo/redo is used.
+	const groupedElements = Object.entries(
+		editableElements
+			.sort( ( a, b ) => ( attributes[ a.props.id ].order as number ) - ( attributes[ b.props.id ].order as number ) )
+			.reduce( ( acc: Record<string, Array<JSX.Element>>, element ) => {
+				const row = attributes[ element.props.id ].row as string;
+				acc[ row ] = acc[ row ] || [];
+				acc[ row ].push( element );
+
+				return acc;
+			}, {} )
+	);
+
+	return (
+		<>
+			<h2 className="subtitle">Multi-root Editor Demo (rich integration)</h2>
+			<p className="info">
+				This sample demonstrates the more advanced integration of a multi-root editor in React.<br/>
+				There are buttons to add/remove roots in the editor, illustrating how this functionality could be implemented.<br/>
+				Additionally, the &apos;Simulate an error&apos; button is used to trigger an error, causing a restart of the editor
+				by Watchdog. Watchdog can be disabled by passing the `disableWatchdog` flag to the `useMultiRootEditor` hook.<br/>
+				There is also an implementation of the &apos;Toggle read-only&apos; button to demonstrate the functionality of changing
+				the read-only mode for the selected editor root.
```

Comment:
```suggestion
				This sample demonstrates a more advanced integration of the multi-root editor in React.<br/>
				Multiple extra features were implemented to illustrate how you can customize your application and use the provided API.<br/>
				They are optional, and you do not need to include them in your application. However, they can be a good starting point for your own custom features.<br/>
				The "Simulate an error" button makes the editor throw an error to show you how it is restarted by the `Watchdog` mechanism. Note, that this mechanism is available by default and you don't need to write additional code to use it. It can be disabled by passing the `disableWatchdog` flag to the `useMultiRootEditor` hook.
```